src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,142 @@
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> nodeAtIndex = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(nodeAtIndex.prev, value, nodeAtIndex);
+        if (nodeAtIndex.prev != null) {
+            nodeAtIndex.prev.next = newNode;
+        } else {
+            head = newNode;
+        }
+        nodeAtIndex.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T lists : list) {
+            add(lists);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T setNode = findNodeByIndex(index).item;
+        findNodeByIndex(index).item = value;
+        return setNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removeNode = findNodeByIndex(index);
+        unlink(removeNode);
+        return removeNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> nodeRemove = head;
+        while (nodeRemove != null) {
+            if ((nodeRemove.item == object)
+                    || (nodeRemove.item != null && nodeRemove.item.equals(object))) {
+                unlink(nodeRemove);
+                return true;
+            }
+            nodeRemove = nodeRemove.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("The index "
+                    + index + " is out of bounds");
+        }
+
+        Node<T> indexNode;
+        if (index < size / 2) {
+            indexNode = head;
+            for (int i = 0; i < index; i++) {
+                indexNode = indexNode.next;
+            }
+        } else {
+            indexNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                indexNode = indexNode.prev;
+            }
+        }
+        return indexNode;
+    }
+
+    public void unlink(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+        size--;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("The index "
+                    + index + " is out of bounds");
```

Comment:
include size in the message, as it is in typical ArrayIndexOutOfBounds exception